// rut-chileno/esm2020/lib/rut.module.mjs
import { NgModule } from '@angular/core';
import { RutComponent } from './rut.component';
import { FormsModule } from '@angular/forms';
import * as i0 from "@angular/core";
export class RutModule {
}
RutModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.11", ngImport: i0, type: RutModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
RutModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.11", ngImport: i0, type: RutModule, declarations: [RutComponent], imports: [FormsModule], exports: [RutComponent] });
RutModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.11", ngImport: i0, type: RutModule, imports: [[
            FormsModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.11", ngImport: i0, type: RutModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        RutComponent
                    ],
                    imports: [
                        FormsModule
                    ],
                    exports: [
                        RutComponent
                    ]
                }]
        }] });
